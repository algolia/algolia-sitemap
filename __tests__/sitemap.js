/* eslint-env jest */
const { createSitemapindex, createSitemap } = require('../sitemap');

describe('sitemap', () => {
  it('renders empty if needed', () => {
    const sitemap = createSitemap();
    expect(sitemap.stringify()).toMatchSnapshot();
  });

  it('renders with a single loc-only entry', () => {
    const entries = [{ loc: 'https://www.example.com' }];
    const sitemap = createSitemap(entries);
    expect(sitemap.stringify()).toMatchSnapshot();
  });

  it('renders with multiple loc-only entries', () => {
    const entries = [
      { loc: 'https://www.example.com' },
      { loc: 'https://www.example.com/test' },
    ];
    const sitemap = createSitemap(entries);
    expect(sitemap.stringify()).toMatchSnapshot();
  });

  it('renders with multiple optional entries', () => {
    const entries = [
      {
        loc: 'https://www.example.com',
        lastmod: new Date(1).toISOString(),
        changefreq: 'daily',
        priority: 0,
      },
      {
        loc: 'https://www.example.com/test',
        lastmod: new Date(0).toISOString(),
        changefreq: 'weekly',
        priority: 0.2,
      },
    ];
    const sitemap = createSitemap(entries);
    expect(sitemap.stringify()).toMatchSnapshot();
  });

  describe('Validation', () => {
    it('ensures loc is required', () => {
      [
        {},
        {
          changefreq: 'weekly',
        },
        {
          alternates: [],
        },
      ].forEach(entry => {
        expect(() => {
          createSitemap([entry]);
        }).toThrow();
      });
    });

    it('ensures lastmod is a w3c date', () => {
      [
        { loc: 'https://example.com', lastmod: '' },
        { loc: 'https://example.com', lastmod: '15 august' },
        { loc: 'https://example.com', lastmod: new Date(1) },
        { loc: 'https://example.com', lastmod: new Date(1).toString() },
        { loc: 'https://example.com', lastmod: new Date(1).toDateString() },
      ].forEach(entry => {
        expect(() => {
          createSitemap([entry]);
        }).toThrow();
      });
    });

    it('ensures priority is between 0 and 1', () => {
      [
        {
          loc: 'https://www.example.com/test',
          priority: NaN,
        },
        {
          loc: 'https://www.example.com/test',
          priority: '0',
        },
        {
          loc: 'https://www.example.com/test',
          priority: -5,
        },
        {
          loc: 'https://www.example.com/test',
          priority: 1.1,
        },
      ].forEach(entry => {
        expect(() => {
          createSitemap([entry]);
        }).toThrow();
      });
    });

    it('ensures changefreq follows spec', () => {
      [
        {
          loc: 'https://www.example.com/test',
          changefreq: 'WEEKLY',
        },
        {
          loc: 'https://www.example.com/test',
          changefreq: '0',
        },
        {
          loc: 'https://www.example.com/test',
          changefreq: null,
        },
      ].forEach(entry => {
        expect(() => {
          createSitemap([entry]);
        }).toThrow();
      });
    });
  });

  it('renders with alternatives', () => {
    const entries = [
      {
        loc: 'https://www.example.com',
        alternates: {
          languages: ['en', 'fr'],
          hitToURL: lang => `https://${lang}.example.com`,
        },
      },
    ];
    const sitemap = createSitemap(entries);
    expect(sitemap.stringify()).toMatchSnapshot();
  });
});

describe('sitemapindex', () => {
  it('renders empty if needed', () => {
    const index = createSitemapindex();
    expect(index.stringify()).toMatchSnapshot();
  });

  it('renders with a single entry', () => {
    const sitemaps = [{ loc: 'https://example.com/sitemap.0.xml' }];
    const index = createSitemapindex(sitemaps);
    expect(index.stringify()).toMatchSnapshot();
  });

  it('renders with multiple entries', () => {
    const sitemaps = [
      { loc: 'https://example.com/sitemap.0.xml' },
      { loc: 'https://example.com/sitemap.1.xml' },
    ];
    const index = createSitemapindex(sitemaps);
    expect(index.stringify()).toMatchSnapshot();
  });
});
