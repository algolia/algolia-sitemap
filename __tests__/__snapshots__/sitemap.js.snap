// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sitemap Validation ensures alternates are returning urls 1`] = `
"alternates {} was not valid. An object with

{
  languages: ['nl-BE', 'fr', 'pt-BR'],
  hitToURL: (language) => \`the url for that \${language}\`,
}

was expected."
`;

exports[`sitemap Validation ensures alternates are returning urls 2`] = `
"alternates {\\"languages\\":\\"\\"} was not valid. An object with

{
  languages: ['nl-BE', 'fr', 'pt-BR'],
  hitToURL: (language) => \`the url for that \${language}\`,
}

was expected."
`;

exports[`sitemap Validation ensures alternates are returning urls 3`] = `
"alternates {\\"languages\\":[\\"\\"]} was not valid. An object with

{
  languages: ['nl-BE', 'fr', 'pt-BR'],
  hitToURL: (language) => \`the url for that \${language}\`,
}

was expected."
`;

exports[`sitemap Validation ensures changefreq follows spec 1`] = `
"changefreq \\"WEEKLY\\" was not a valid value.

expected: always, hourly, daily, weekly, monthly, yearly, never

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures changefreq follows spec 2`] = `
"changefreq \\"0\\" was not a valid value.

expected: always, hourly, daily, weekly, monthly, yearly, never

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures changefreq follows spec 3`] = `
"changefreq \\"null\\" was not a valid value.

expected: always, hourly, daily, weekly, monthly, yearly, never

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures lastmod is a w3c date 1`] = `
"lastmod \\"\\" is not valid. It should be a valid ISO 8601 date

Try using new Date().toISOString()

see https://www.sitemaps.org/protocol.html
see https://www.w3.org/TR/NOTE-datetime"
`;

exports[`sitemap Validation ensures lastmod is a w3c date 2`] = `
"lastmod \\"15 august\\" is not valid. It should be a valid ISO 8601 date

Try using new Date().toISOString()

see https://www.sitemaps.org/protocol.html
see https://www.w3.org/TR/NOTE-datetime"
`;

exports[`sitemap Validation ensures lastmod is a w3c date 3`] = `
"lastmod \\"Thu Jan 01 1970 00:00:00 GMT+0000 (UTC)\\" is not valid. It should be a valid ISO 8601 date

Try using new Date().toISOString()

see https://www.sitemaps.org/protocol.html
see https://www.w3.org/TR/NOTE-datetime"
`;

exports[`sitemap Validation ensures lastmod is a w3c date 4`] = `
"lastmod \\"Thu Jan 01 1970\\" is not valid. It should be a valid ISO 8601 date

Try using new Date().toISOString()

see https://www.sitemaps.org/protocol.html
see https://www.w3.org/TR/NOTE-datetime"
`;

exports[`sitemap Validation ensures loc is required 1`] = `
"loc \\"undefined\\" was not valid. It's required.

see https://www.sitemaps.org/protocol.html"
`;

exports[`sitemap Validation ensures loc is required 2`] = `
"loc \\"undefined\\" was not valid. It's required.

see https://www.sitemaps.org/protocol.html"
`;

exports[`sitemap Validation ensures loc is required 3`] = `
"loc \\"undefined\\" was not valid. It's required.

see https://www.sitemaps.org/protocol.html"
`;

exports[`sitemap Validation ensures priority is between 0 and 1 1`] = `
"
priority \\"NaN\\" was not valid. A number between 0 and 1 is expected.

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures priority is between 0 and 1 2`] = `
"
priority \\"0\\" was not valid. A number between 0 and 1 is expected.

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures priority is between 0 and 1 3`] = `
"
priority \\"-5\\" was not valid. A number between 0 and 1 is expected.

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap Validation ensures priority is between 0 and 1 4`] = `
"
priority \\"1.1\\" was not valid. A number between 0 and 1 is expected.

see https://www.sitemaps.org/protocol.html for more information"
`;

exports[`sitemap renders empty if needed 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
</urlset>

`;

exports[`sitemap renders with a single loc-only entry 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>
      https://www.example.com
    </loc>
  </url>
</urlset>

`;

exports[`sitemap renders with alternatives 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url xmlns:xhtml="http://www.w3.org/1999/xhtml">
    <loc>
      https://www.example.com
    </loc>
    <xhtml:link rel="alternate"
                hreflang="en"
                href="https://en.example.com"
    >
    </xhtml:link>
    <xhtml:link rel="alternate"
                hreflang="fr"
                href="https://fr.example.com"
    >
    </xhtml:link>
  </url>
</urlset>

`;

exports[`sitemap renders with multiple loc-only entries 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>
      https://www.example.com
    </loc>
  </url>
  <url>
    <loc>
      https://www.example.com/test
    </loc>
  </url>
</urlset>

`;

exports[`sitemap renders with multiple optional entries 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>
      https://www.example.com
    </loc>
    <lastmod>
      1970-01-01T00:00:00.001Z
    </lastmod>
    <changefreq>
      daily
    </changefreq>
    <priority>
      0.0
    </priority>
  </url>
  <url>
    <loc>
      https://www.example.com/test
    </loc>
    <lastmod>
      1970-01-01T00:00:00.000Z
    </lastmod>
    <changefreq>
      weekly
    </changefreq>
    <priority>
      0.2
    </priority>
  </url>
</urlset>

`;

exports[`sitemapindex renders empty if needed 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
</sitemapindex>

`;

exports[`sitemapindex renders with a single entry 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <sitemap>
    <loc>
      https://example.com/sitemap.0.xml
    </loc>
  </sitemap>
</sitemapindex>

`;

exports[`sitemapindex renders with multiple entries 1`] = `

<?xml version="1.0"
      encoding="utf-8"
?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <sitemap>
    <loc>
      https://example.com/sitemap.0.xml
    </loc>
  </sitemap>
  <sitemap>
    <loc>
      https://example.com/sitemap.1.xml
    </loc>
  </sitemap>
</sitemapindex>

`;
